File: Daily Challenge/Java/Longest Common Substring 30.3.2025.java
import java.util.*;

public class Main {
    public static List<String> getSymmetricSubstrings(String s, int m) {
        List<String> substrings = new ArrayList<>();
        for (int ln = Math.min(s.length() - m - 1, m) * 2 + 1; ln > 0; ln--) {
            int h = ln / 2;
            if (h >= 0) {
                substrings.add(s.substring(m - h, m + h + 1));
            }
        }
        return substrings;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String a = sc.nextLine().trim();
        String b = sc.nextLine().trim();
        sc.close();

        if (a.isEmpty() || b.isEmpty()) {
            System.out.println(-1);
            return;
        }

        int c = a.length() / 2, d = b.length() / 2;
        List<String> e = getSymmetricSubstrings(a, c);
        Set<String> f = new HashSet<>(getSymmetricSubstrings(b, d));

        for (String sub : e) {
            if (f.contains(sub)) {
                System.out.println(sub);
                return;
            }
        }
        System.out.println(-1);
    }
}

File: Daily Challenge/C++/Longest Common Substring 30.3.2025.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

vector<string> getSymmetricSubstrings(string s, int m) {
    vector<string> substrings;
    int len = s.length();
    for (int ln = min(len - m - 1, m) * 2 + 1; ln > 0; ln--) {
        int h = ln / 2;
        if (h >= 0) {
            substrings.push_back(s.substr(m - h, ln));
        }
    }
    return substrings;
}

int main() {
    string a, b;
    getline(cin, a);
    getline(cin, b);

    if (a.empty() || b.empty()) {
        cout << "-1\n";
        return 0;
    }

    int c = a.length() / 2, d = b.length() / 2;
    vector<string> e = getSymmetricSubstrings(a, c);
    unordered_set<string> f(getSymmetricSubstrings(b, d).begin(), getSymmetricSubstrings(b, d).end());

    for (string sub : e) {
        if (f.find(sub) != f.end()) {
            cout << sub << "\n";
            return 0;
        }
    }

    cout << "-1\n";
    return 0;
}
